pkg/cri/opts/container.go

```diff
@@ -97,8 +97,22 @@ func WithVolumes(volumeMounts map[string]string) containerd.NewContainerOpts {
 		}()
 
 		for host, volume := range volumeMounts {
-			// The volume may have been defined with a C: prefix, which we can't use here.
-			volume = strings.TrimPrefix(volume, "C:")
+			// Windows allows volume mounts in subfolders under C: and as any other drive letter like D:, E:, etc.
+			// An image may contain files inside a folder defined as a VOLUME in a Dockerfile. On Windows, images
+			// can only contain pre-existing files for volumes situated on the root filesystem, which is C:.
+			// For any other volumes, we need to skip attempting to copy existing contents.
+			//
+			// C:\some\volume --> \some\volume
+			// D:\some\volume --> skip
+			if len(volume) >= 2 && string(volume[1]) == ":" {
+				// Perform a case insensitive comparison to "C", and skip non-C mounted volumes.
+				if !strings.EqualFold(string(volume[0]), "c") {
+					continue
```

Comment:
I am not a windows fan, is this something a user knows in advance? do we need to add a log?

Reply:
> I am not a windows fan

That's quite alright. I am happy to provide some more context.  

> is this something a user knows in advance?

It's somewhat documented here: https://docs.docker.com/engine/reference/commandline/run/#volume

> do we need to add a log?

TL;DR

I don't think a log message here would be helpful. There is nothing to correct and the user can't really do anything about the situation. I will explain bellow.

TS;NM

The code here tries to populate a ```VOLUME``` with pre-existing files from within an image. For any volume mounted somewhere outside the root filesystem, we can't really store files within an image. At least not without a spec change (I may be wrong, need to check).

On Linux, all volumes get mounted somewhere under ```/```. So if we try to mount a docker volume, a block volume or a remote filesystem like NFS, all mounts end up somewhere under ```/```. The DOS drive letters in Windows are like having multiple ```/``` on Linux, but only one of which is the actual boot-able filesystem. Windows allows us to mount filesystem inside folders, just like on Linux, but it also allows us to mount filesystems as different drive letters, which as I mentioned, act like a separate ```/```.

So in essence, for pre-existing ```VOLUME``` contents we can easily support the case where volumes are mounted as folders (which is identical to Linux), but we can't support the case where volumes are mounted as different drive letters, as the image build process doesn't capture files that are not under ```C:```. 

For the case where there is a ```VOLUME``` mounted as a different drive letter, the target must always be a bare drive letter like: ```D:``` and not a sub-folder of that drive letter like: ```D:\``` or ```D:\somepath```. This is because we must first have a valid store for a drive letter and that drive letter set up, before we can mount anything under it. It's like first having to mount a FS under ```/``` on Linux before you can mount anything in ```/var```.

Reply:
Thanks for telling me why it is done.

I wasn't sure if users would be confused after skipping some volume copies and how we could make this easier for them to know. Of course if this is common knowledge in windows, then I guess my fears are superfluous :grin:

Reply:
It's also implicit from the [`VOLUME` documentation](https://docs.docker.com/engine/reference/builder/#notes-about-specifying-volumes), because the only way to have a non-`C:` mount on Windows during build-time is by calling `VOLUME ["D:"]` earlier (which will as a new volume, will be empty), and as noted, anything done to that volume after that step is discarded, same as Linux. (In the future, BuildKit will support `RUN --mount` which could also create non-C: volumes, but again, any changes would not be persisted to the image, consistent with Linux).

Reply:
I don't think this works for linux volumes where `:` can be a valid char for a directory path and whether or not the pattern is `a-Z:` the pair of chars are not considered to be a drive letter..

granted it may be "convention" to not try to use `a-Z:` as a prefix on a linux volume.. still not sure we need to force this limitation on linux.. hmm

noting `a-Z` on windows is `A-Z` (to upper) after to upper since windows is case insensitive :-) 

I would feel much better about this if this code path for stripping and/or skipping volumes was limited to windows build versions of containerd.

Reply:
I would feel much better about this if this code path for stripping and/or skipping volumes was limited to windows build versions of containerd.

Reply:
nod /: was in the back of my mind... :)

Reply:
perhaps build goos && do a check in the windows build if wcow

Reply:
the Stat check should be sufficient otherwise..

Reply:
A few notes:

  * On Linux, It is possible to create volumes that are not absolute: ```VOLUME ["test"]```. They will be relative to the root filesystem. On Windows, docker throws ```Unrecognised volume spec: invalid volume specification: 'test'```.
  * On linux, the current code has the same issue as discussed here. Paths containing ```:``` are valid. Thus, so is a volume defined as: ```VOLUME ["C:/test"]```.
  * The current code strips away ```C:``` which mangles the user defined ```VOLUME```. With the current ```strings.TrimPrefix(volume, "C:")``` removed, on linux we get an image with a volume mounted under: ```/C:/test```, which seems to be valid:

```bash
/ # ls
C:     bin    dev    etc    home   lib    lib64  proc   root   sys    tmp    usr    var
/ # ls /C:/test/
test.txt
```

We have a few options to consider:

  * Disallow drive letter notation by convention
  * Put this check behind a ```runtime.GOOS``` check, but we break LCOW. Containerd doesn't (yet?) support LCOW from what I see, but this would make it even harder to support in the future.
  * Mutate the volumes before we pass them into ```WithVolumes()```
  * Make the container target platform available to ```WithVolumes()```. This would require either changing the signature of the function or adding a new field to the spec, which must be set before we call ```WithVolumes()```

What do you folks think would be the best path forward?

Reply:
`WithVolumes` is only called from two places, basically identically, so I'd add "`image.ImageSpec.OS`" to that call. I had been assuming that the relevant information would be in `containers.Container` but on consideration it's probably hidden behind the Runtime choice.

That is both the correct check (since it's the image that we're making volumes out of) should be sufficient, and avoids a build/goos check, just in case someone implements a VM-type runtime that can execute Windows containers from a Linux-hosted containerd.

Reply:
The target platform does get added to the ```Spec```, but it's done later and it's marshaled. There is no easy way to get the platform without moving code around and unmarshaling that ```typeurl.Any``` field, which would end up being an ugly hack at best. I'd rather not traumatize the next person that would have to look at that code in the future.

@mikebrow @Iceber @dcantah @thaJeztah Does adding a new parameter to ```WithVolumes()``` indicating the target platform sound good to you? Do you have other suggestions? To properly make a decision in the code, we do need that info made available to ```WithVolumes()```.

Reply:
I made the change in a separate commit. PTAL.

Reply:
> the Stat check should be sufficient otherwise..

On Windows, if we don't strip the drive letter, ```fs.RootPath()``` will try to join the root path on the host, to the volume defined by the user, and we end up with ```C:\ProgramData\containerd\....truncated....\D:``` which is an invalid path on Windows and throws an error. The ```os.Stat()``` check will fail with an error that does not match ```os.IsNotExist()```.

Reply:
sigh.. I meant if all this was hidden under a .windows.go  then I wouldn't have to worry about it :)

Reply:
I would have loved to hide this under a ```_windows.go```, but Windows hosts can also create Linux containers :sweat_smile:. This code path needs to be applied to Windows Containers On Windows (```WCOW```) and skipped on Linux Containers On Windows (```LCOW```).

Reply:
WithVolumes change looks great, trying to think of any esoteric edge cases though..